cockpit/leapp.html

```diff
@@ -267,29 +292,142 @@
+                }
+            }
+
+            return mapped_ports;
         }
 
-        function show_target_ports() {
-            // maybe it is a good place to unlock migrate button?
+        var ports_scan_range = [];
+        function check_target_ports() {
             if (! $("input[name=target-machine]:checked").length ||
                 ! $("input[name=source-machine]:checked").length) {
                 return;
             }
-            $("#target-ports ul li").remove();
+
+            var ip_addr = $("input[name=target-machine]:checked").data('ip');
+
+            var selected_ports_cnt = $("#panel-body input.source-ports:checked").length;
+            var min_port = null;
+            var max_port = null;
+            var ports_interval = 25;
+
+            function add_ports_range(min_port, max_port) {
+                ports_scan_range.push(min_port.toString() + '-' + max_port.toString());
+            }
+
+            $("#panel-body input.source-ports:checked").map(function(index) {
+                selected_ports_cnt--;
+
+                if (min_port === null) {
+                    min_port = parseInt(this.value);
+                    max_port = min_port;
+                }
+
+                if (parseInt(this.value) <= max_port) {
+                    max_port += ports_interval;
+
+                    if (selected_ports_cnt <= 0) {
+                        add_ports_range(min_port, max_port);
+                    }
+
+                } else {
+                    // get element from last loop
+                    add_ports_range(min_port, max_port);
+
+                    min_port = parseInt(this.value);
+                    max_port = min_port + ports_interval;
+
+                    if (selected_ports_cnt <= 0) {
+                        add_ports_range(min_port, max_port);
+                    }
+                }
+
+            });
+
+            if (ports_scan_range) {
+                scan_ports(ports_scan_range.join(','), ip_addr, show_target_ports);
+            } else {
+                console.log("couldn not find selected ports");
+            }
+        }
+
+        function show_target_ports(data) {
+            data = JSON.parse(data);
+
             loader.show();
             output.append(document.createTextNode("Mapping target ports"));
-            $("#panel-body input[name=source-ports]:checked").each(function() {
+
+            var selected_ports = $("#panel-body input.source-ports:checked").map(function() {
+                return this.value;
+            }).toArray();
+
+            if (typeof(data.ports.tcp) !== undefined && data.ports.tcp) {
+                var busy_ports = data.ports.tcp  
+            } else {
+                var busy_ports = {};
+            };
+
+            var mapped_ports = map_ports(ports_scan_range, busy_ports, selected_ports);
+
+            $("#target-ports ul li").remove();
+            for (target_port in mapped_ports) {
```

Comment:
mapped_ports is an object, it's not an array, when it comes to code above, should be:
```
Object.keys(mapped_ports).forEach...
```
so I think it's better to use `in`